### DISCUSSION TITLE
How to implement channel-specific MIDI gate, gain and freq?

### DISCUSSION THREAD

**Discussion by oshibka404** [grame-cncm/faust#696] [Q&A]
Is there a way to make channel-specific `freq`, `gain` and `gate` parameters?
Like, if I wanted to have a voice for bass controlled by MIDI channel 1 and a voice for lead synth controlled by MIDI channel 2 and wrote something like this:
```
bass_gate = button("Bass Gate[midi:gate 1]");
bass_gain = hslider("Bass Gain[midi:gain 1]");
bass_freq = hslider("Bass Freq[midi:freq 1]");

lead_gate = button("Lead Gate[midi:gate 2]");
lead_gain = hslider("Lead Gain[midi:gain 2]");
lead_freq = hslider("Lead Freq[midi:freq 2]");
```
and press of a MIDI key on channel 1 would affect `bass_gate`, `bass_gain` and `bass_freq` but not others.

I've read the [MIDI Manual](https://faustdoc.grame.fr/manual/midi/) and it seems like I can have very advanced channel-specific CC controls and even values assigned to the velocity of particular MIDI keys on particular channels (with aftertouch!), but not the basic freq-gain-gate trio. Is it true or I am missing something?

Thanks.

**Comment by sletz** [comment]
The current [mydsp_poly](https://github.com/grame-cncm/faust/blob/8b86f3df423cfb4e781e106eb88530f317267198/architecture/faust/dsp/poly-dsp.h#L512) has `keyOn/keyOff` methods with a channel parameter which is not used for now.  We could possibly expand the model to handle the channel parameter and create a multi-channel aware version with an array of 16 `mydsp_poly` instruments, and dispatch on each of then using the channel. Or only creating the instruments for the needed channels?

**Reply by goofy2k** [reply]
I believe that this is more or less what I want to achieve by setting voice parameters per voice, based on the channel as input.... see issue #695

**Comment by goofy2k** [comment]
I use faust2api for ESP32 and want to play notes with channel-specific DSP parameters. I have a solution, which is in my opinion pretty "dirty", but it works fine for me now. 

For now I designed flexible keyOn routine. This starts playing a note, but immediately shuts it down, by setting the gate parameter to zero. Then I load a predefined and channel-dependent set of parameters using setVoiceParamValue commands. After the desired parameters have been load into the DSP engine I fire the note again by setting gate to 1.

[flex_keyOn_cpp.txt](https://github.com/grame-cncm/faust/files/7900985/flex_keyOn_cpp.txt)

I have to use this dirty method as the API only (if I am correct!) allows setting a voice specific parameters when the note has already been started.   The alternative is getting the voice address with newVoice and then play it by setting the gate to 1. But then you have to manage the active voices yourself. Again, if I am correct.....

**Comment by sletz** [comment]
One solution could be extending the [process + effect](https://faustdoc.grame.fr/manual/midi/#audio-effects-and-polyphonic-synthesizer) model: 
- have a new convention like `process1 = exp; .... process128 = exp` to define up to 128 instruments and `effect1 = exp; .... effect128 = exp` their possible associated global effects
- have the C++ architecture file extract each of `processN/effectN` DSP, compile them separately, combine them at C++ level
- then use MIDI prog change events to associate an instrument with a given MIDI channel

**Comment by pkamenarsky** [comment]
Has there been any progress on this?

**Reply by akjmicro** [reply]
Yes, I actually asked precisely this on the mailing list; glad others are wondering too. Seems a shame not to have sensible MIDI channel isolation like any other DPS package would...
